test/sys/test_socket.rs

```diff
@@ -220,12 +223,84 @@ mod recvfrom {
             SockFlag::empty(),
             None,
         ).expect("send socket failed");
-        let from = sendrecv(rsock, ssock, move |s, m, flags| {
+        let from = sendrecv(rsock, ssock, None, move |s, m, flags| {
             sendto(s, m, &sock_addr, flags)
         });
         // UDP sockets should set the from address
         assert_eq!(AddressFamily::Inet, from.unwrap().family());
     }
+
+    #[cfg(target_os = "linux")]
+    mod udp_offload {
+        use super::*;
+        use nix::sys::uio::IoVec;
+        use nix::sys::socket::sockopt::{UdpGroSegment, UdpGsoSegment};
+
+        #[test]
+        pub fn gso() {
+            require_kernel_version!(udp_offload::gso, ">= 4.18");
+
+            let std_sa = SocketAddr::from_str("127.0.0.1:6791").unwrap();
+            let inet_addr = InetAddr::from_std(&std_sa);
+            let sock_addr = SockAddr::new_inet(inet_addr);
+            let rsock = socket(AddressFamily::Inet,
+                               SockType::Datagram,
+                               SockFlag::empty(),
+                               None
+            ).unwrap();
+
+            setsockopt(rsock, UdpGsoSegment, &2).expect("setsockopt UDP_SEGMENT failed");
+
+            bind(rsock, &sock_addr).unwrap();
+            let ssock = socket(
+                AddressFamily::Inet,
+                SockType::Datagram,
+                SockFlag::empty(),
+                None,
+            ).expect("send socket failed");
+
+            let from = sendrecv(rsock, ssock, Some(2), move |s, m, flags| {
+                let iov = [IoVec::from_slice(m)];
+                let cmsg = ControlMessage::UdpGsoSegments(&2);
+                sendmsg(s, &iov, &[cmsg], flags, Some(&sock_addr))
+            });
+            // UDP sockets should set the from address
+            assert_eq!(AddressFamily::Inet, from.unwrap().family());
+        }
+
+        #[test]
+        pub fn gro() {
```

Comment:
How does this test work?  I don't understand what effect the UDP_GRO socket option has, or what role the ControlMessage::UdpGsoSegments has.

Reply:
I've lost the comment with the explanation. The problem is that it's not clear how to guarantee that incoming packets are GROed. It may vary between kernel versions, and it may just be impossible on a physical network interface. I've seen some GRO tests working on veth interfaces. I will need some time to research. This test is just setting the socket option and check that it doesn't fail with an error.

Reply:
If it's truly unobservable then there's no point in writing a test that doesn't do anything.  Just make sure that you can set the socket options without error.  There are a few functions like that in Nix.